### PR TITLE
[ZEN2] Use Zen2 in REST Tests

### PR DESCRIPTION
**Pull request by original-brownbear**

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java [hunk 1]
```diff
@@ -72,15 +72,17 @@ public void start() {
...
+        if ((initialMasterNodeCount > 0 || initialMasterNodes.isEmpty() == false) && transportService.getLocalNode().isMasterNode()) {
             logger.debug("unsafely waiting for discovery of [{}] master-eligible nodes", initialMasterNodeCount);
 
             final ThreadContext threadContext = transportService.getThreadPool().getThreadContext();
             try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
                 threadContext.markAsSystemContext();
 
                 final GetDiscoveredNodesRequest request = new GetDiscoveredNodesRequest();
-                request.setWaitForNodes(initialMasterNodeCount);
+                if (initialMasterNodeCount > 0) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java]: @DaveCTurner I think there's only a UT for this case left here. Before I add one: Is this ok? Simply leaving the default of `1` for the `initalMasterNodeCount` on the request if we have the list?

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy [hunk 2]
```diff
@@ -134,6 +136,13 @@ class ClusterFormationTasks {
                         esConfig['discovery.zen.hosts_provider'] = 'file'
                     }
                     esConfig['discovery.zen.ping.unicast.hosts'] = []
+                    if (esConfig.containsKey('discovery.type') == false
+                      && config.minimumMasterNodes.call() > 1 && hasBwcNodes == false) {
```
Comment (DaveCTurner) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy]: Could this condition just be `hasBwcNodes == false`?

File: qa/full-cluster-restart/build.gradle [hunk 3]
```diff
@@ -71,6 +71,8 @@ for (Version version : bwcVersions.indexCompatible) {
 
     // some tests rely on the translog not being flushed
     setting 'indices.memory.shard_inactive_time', '20m'
+    // TODO: Use zen2 in bwc tests
+    setting 'discovery.type', 'zen'
```
Comment (DaveCTurner) [inline qa/full-cluster-restart/build.gradle]: I _think_ these tests should work with Zen2 already. Did you try and find them not to? If not, could you try?

File: server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java]: I think we should set both properties (for now).

File: x-pack/qa/full-cluster-restart/build.gradle [hunk 4]
```diff
@@ -216,6 +216,8 @@ subprojects {
           setting 'xpack.watcher.encrypt_sensitive_data', 'true'
           keystoreFile 'xpack.watcher.encryption_key', "${mainProject.projectDir}/src/test/resources/system_key"
       }
+      // TODO: Use zen2 in bwc tests
+      setting 'discovery.type', 'zen'
```
Comment (DaveCTurner) [inline x-pack/qa/full-cluster-restart/build.gradle]: I _think_ this test should work with Zen2.

**Review by DaveCTurner — COMMENTED** [review]
Great. I left a few questions.

File: qa/full-cluster-restart/build.gradle (hunk 3 above)
Comment (original-brownbear) [reply qa/full-cluster-restart/build.gradle]: I tried these manually and they failed. It seems the metadata is not read correctly.

I'm seeing 

```
[2018-12-06T10:58:30,663][INFO ][o.e.c.c.Coordinator      ] [node-0] setting initial configuration to VotingConfiguration{jxnfx7BsSbuwcz0S8v3psg,soe4_OeBTSGNJm66mOyu0g}
[2018-12-06T10:58:30,803][INFO ][o.e.c.s.MasterService    ] [node-0] elected-as-master ([2] nodes joined)[{node-1}{soe4_OeBTSGNJm66mOyu0g}{c7jlFdFLSiumcLFCWqNjzw}{127.0.0.1}{127.0.0.1:45245}{testattr=test} elect leader, {node-0}{jxnfx7BsSbuwcz0S8v3psg}{hhlkfDtYTE2hZeUKtYFYXg}{127.0.0.1}{127.0.0.1:46175}{testattr=test} elect leader, _BECOME_MASTER_TASK_, _FINISH_ELECTION_], reason: master node changed {previous [], current [{node-0}{jxnfx7BsSbuwcz0S8v3psg}{hhlkfDtYTE2hZeUKtYFYXg}{127.0.0.1}{127.0.0.1:46175}{testattr=test}]}, added {{node-1}{soe4_OeBTSGNJm66mOyu0g}{c7jlFdFLSiumcLFCWqNjzw}{127.0.0.1}{127.0.0.1:45245}{testattr=test},}
[2018-12-06T10:58:30,866][INFO ][o.e.c.s.ClusterApplierService] [node-0] master node changed {previous [], current [{node-0}{jxnfx7BsSbuwcz0S8v3psg}{hhlkfDtYTE2hZeUKtYFYXg}{127.0.0.1}{127.0.0.1:46175}{testattr=test}]}, added {{node-1}{soe4_OeBTSGNJm66mOyu0g}{c7jlFdFLSiumcLFCWqNjzw}{127.0.0.1}{127.0.0.1:45245}{testattr=test},}, reason: Publication{term=1, version=2}
[2018-12-06T10:58:31,048][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testhistoryuuidisadded/FNWuN8mXSy6vqEFEU4RkHQ]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,048][INFO ][o.e.g.DanglingIndicesState] [node-0] [[info/v9CnNlBoQrOFSmx6mKMe9Q]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,048][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testsingledoc/NlUIRH7xSSaT1SECBfMmOA]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testrecovery/OsAKBDv0SmybzcNvsxCLgw]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testrollover-000001/31YMrsVzSlSaZf0ODLYoiA]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testrollover-000002/L1hhzmKJT8SW27-ghSvU2Q]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testshrink/uZHLza9TSAu_U64zfzhRsQ]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[test_empty_shard/NCCsphB_SvC7pftlKcoegw]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testnewreplicaswork/i1xgSMENTvqDMDCcFR7FeQ]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testsearch/UWA9LkzgTnqpxwhM1aMiNg]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testsnapshotrestore/1RZJMl1-RumJIJ9BKigC4g]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testshrinkafterupgrade/QbjmPie_SLyQWFxUyKlbxg]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,049][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testclusterstate/W4RYzvS_TFqNYfNG1VCR_A]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,050][INFO ][o.e.g.DanglingIndicesState] [node-0] [[queries/gFrWR58uQp6oBH9NWZ2clQ]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,050][INFO ][o.e.g.DanglingIndicesState] [node-0] [[testshrink_shrunk/VdjBr7yOTYKcvYXZzsv1lQ]] dangling index exists on local file system, but not in cluster metadata, auto import to cluster state
[2018-12-06T10:58:31,054][INFO ][o.e.g.GatewayService     ] [node-0] recovered [0] indices into cluster_state
=========================================
```

and failing tests on the fully upgraded cluster step.

File: x-pack/qa/full-cluster-restart/build.gradle (hunk 4 above)
Comment (original-brownbear) [reply x-pack/qa/full-cluster-restart/build.gradle]: Same answer as here https://github.com/elastic/elasticsearch/pull/36300#discussion_r239448308.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy (hunk 2 above)
Comment (original-brownbear) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy]: @DaveCTurner I need the first line to have my override of the type to `zen.

But yea the minimum master node check can go away now that the setting of the initial nodes list is there I think, trying it now.

File: qa/full-cluster-restart/build.gradle (hunk 3 above)
Comment (ywelsch) [reply qa/full-cluster-restart/build.gradle]: The problem here is that the bootstrapping (setInitialConfiguration) wipes the original metadata. Need to fix that.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy (hunk 2 above)
Comment (original-brownbear) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy]: Removing the master node count causes at least the docs tests for https://github.com/elastic/elasticsearch/blob/master/docs/reference/modules/node.asciidoc to fail  (rightfully so for zen2 I guess).

Want to fix that here or in a follow up?

File: qa/full-cluster-restart/build.gradle (hunk 3 above)
Comment (DaveCTurner) [reply qa/full-cluster-restart/build.gradle]: I think this fixes it:

```diff
diff --git a/server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java b/server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java
index 4869ace4946..0f2b44bdd92 100644
--- a/server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java
+++ b/server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java
@@ -652,7 +652,7 @@ public class Coordinator extends AbstractLifecycleComponent implements Discovery
                 .lastCommittedConfiguration(votingConfiguration)
                 .build();

-            MetaData.Builder metaDataBuilder = MetaData.builder();
+            MetaData.Builder metaDataBuilder = MetaData.builder(currentState.metaData());
             // automatically generate a UID for the metadata if we need to
             metaDataBuilder.generateClusterUuidIfNeeded(); // TODO generate UUID in bootstrapping tool?
             metaDataBuilder.coordinationMetaData(coordinationMetaData);
```

Can't think of a good way to test this apart from this full-cluster-restart test - there are some that check that settings (e.g. snapshot repos) are preserved across the upgrade, which seems good to me.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy (hunk 2 above)
Comment (original-brownbear) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy]: This test is now skipped :)

File: qa/full-cluster-restart/build.gradle (hunk 3 above)
Comment (original-brownbear) [reply qa/full-cluster-restart/build.gradle]: This fix worked and is part of this PR now.

File: server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java [hunk 5]
```diff
@@ -72,15 +72,17 @@ public void start() {
         assert running == false;
         running = true;
 
-        if (initialMasterNodeCount > 0 && transportService.getLocalNode().isMasterNode()) {
+        if ((initialMasterNodeCount > 0 || initialMasterNodes.isEmpty() == false) && transportService.getLocalNode().isMasterNode()) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java]: This needs changes to the unit tests, I think, something like this perhaps?

```
diff --git a/server/src/test/java/org/elasticsearch/cluster/coordination/ClusterBootstrapServiceTests.java b/server/src/test/java/org/elasticsearch/cluster/coordination/ClusterBootstrapServiceTests.java
index 952d7c0e752..23030b9500f 100644
--- a/server/src/test/java/org/elasticsearch/cluster/coordination/ClusterBootstrapServiceTests.java
+++ b/server/src/test/java/org/elasticsearch/cluster/coordination/ClusterBootstrapServiceTests.java
@@ -48,6 +48,7 @@ import java.util.stream.Stream;
 import static java.util.Collections.emptyMap;
 import static java.util.Collections.emptySet;
 import static java.util.Collections.singleton;
+import static org.elasticsearch.cluster.coordination.ClusterBootstrapService.INITIAL_MASTER_NODES_SETTING;
 import static org.elasticsearch.cluster.coordination.ClusterBootstrapService.INITIAL_MASTER_NODE_COUNT_SETTING;
 import static org.elasticsearch.node.Node.NODE_NAME_SETTING;

@@ -76,8 +77,14 @@ public class ClusterBootstrapServiceTests extends ESTestCase {
         transportService = transport.createTransportService(Settings.EMPTY, deterministicTaskQueue.getThreadPool(),
             TransportService.NOOP_TRANSPORT_INTERCEPTOR, boundTransportAddress -> localNode, null, emptySet());

-        clusterBootstrapService = new ClusterBootstrapService(Settings.builder().put(INITIAL_MASTER_NODE_COUNT_SETTING.getKey(), 3).build(),
-            transportService);
+        final Settings settings;
+        if (randomBoolean()) {
+            settings = Settings.builder().put(INITIAL_MASTER_NODE_COUNT_SETTING.getKey(), 3).build();
+        } else {
+            settings = Settings.builder()
+                .putList(INITIAL_MASTER_NODES_SETTING.getKey(), localNode.getName(), otherNode1.getName(), otherNode2.getName()).build();
+        }
+        clusterBootstrapService = new ClusterBootstrapService(settings, transportService);
     }

     private DiscoveryNode newDiscoveryNode(String nodeName) {
```

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java]: Right, looks good let me try that :)

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java]: The above works fine, added to this PR now :)

**Comment by original-brownbear** [comment]
Jenkins test this

**Comment by ywelsch** [comment]
@elasticmachine retest this please

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**